pkg/liqonet/tunnel/resolver/resolver_test.go

```diff
@@ -0,0 +1,148 @@
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.IPv4(1, 2, 3, 4)},
+					{IP: net.IPv4(1, 2, 3, 5)},
+				}, nil
+			},
+			cache:         map[string]*net.IPAddr{},
+			expectedIP:    "1.2.3.4",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.IPv4(1, 2, 3, 4)}),
+		}),
+
+		Entry("should resolve IPv4 address with cache", &resolveTestcase{
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.IPv4(1, 2, 3, 4)},
+					{IP: net.IPv4(1, 2, 3, 5)},
+				}, nil
+			},
+			cache: map[string]*net.IPAddr{
+				"example.com": {IP: net.IPv4(1, 2, 3, 5)},
+			},
+			expectedIP:    "1.2.3.5",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.IPv4(1, 2, 3, 5)}),
+		}),
+
+		Entry("should resolve IPv6 addresses", &resolveTestcase{
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.ParseIP("2001:db8:a0b:12f0::1")},
+					{IP: net.ParseIP("2001:db8:a0b:12f0::2")},
+				}, nil
+			},
+			cache:         map[string]*net.IPAddr{},
+			expectedIP:    "2001:db8:a0b:12f0::1",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.ParseIP("2001:db8:a0b:12f0::1")}),
+		}),
+
+		Entry("should resolve IPv6 address with cache", &resolveTestcase{
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.ParseIP("2001:db8:a0b:12f0::1")},
+					{IP: net.ParseIP("2001:db8:a0b:12f0::2")},
+				}, nil
+			},
+			cache: map[string]*net.IPAddr{
+				"example.com": {IP: net.ParseIP("2001:db8:a0b:12f0::2")},
+			},
+			expectedIP:    "2001:db8:a0b:12f0::2",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.ParseIP("2001:db8:a0b:12f0::2")}),
+		}),
+
+		Entry("should prefer IPv4 address", &resolveTestcase{
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.ParseIP("2001:db8:a0b:12f0::1")},
+					{IP: net.IPv4(1, 2, 3, 4)},
+				}, nil
+			},
+			cache:         map[string]*net.IPAddr{},
+			expectedIP:    "1.2.3.4",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.IPv4(1, 2, 3, 4)}),
+		}),
+
+		Entry("should prefer IPv4 address with cache", &resolveTestcase{
+			name: "example.com",
+			resolver: func(ctx context.Context, host string) ([]net.IPAddr, error) {
+				return []net.IPAddr{
+					{IP: net.ParseIP("2001:db8:a0b:12f0::1")},
+					{IP: net.IPv4(1, 2, 3, 4)},
+				}, nil
+			},
+			cache: map[string]*net.IPAddr{
+				"example.com": {IP: net.ParseIP("2001:db8:a0b:12f0::1")},
+			},
+			expectedIP:    "1.2.3.4",
+			expectedCache: HaveKeyWithValue("example.com", &net.IPAddr{IP: net.IPv4(1, 2, 3, 4)}),
+		}),
+	)
+
```

Comment:
Could add an entry to check that everything works as expected in case the entry previously in cache is no longer returned during resolution.